File: client-desktop/src/HoYoMusic.Desktop.Core/Contracts/ApiEnvelope.cs
namespace HoYoMusic.Desktop.Core.Contracts;

public sealed class ApiEnvelope<T>
{
    public bool Success { get; init; }
    public T? Data { get; init; }
    public ApiError? Error { get; init; }
}

public sealed class ApiError
{
    public string? Code { get; init; }
    public string? Message { get; init; }
}

File: client-mobile/android/app/src/main/java/com/hoyomusic/mobile/core/network/ApiEnvelope.kt
package com.hoyomusic.mobile.core.network

import com.squareup.moshi.Json

data class ApiEnvelope<T>(
    @Json(name = "success") val success: Boolean,
    @Json(name = "data") val data: T?,
    @Json(name = "error") val error: ApiErrorPayload?
)

data class ApiErrorPayload(
    @Json(name = "code") val code: String?,
    @Json(name = "message") val message: String?
)

